### ISSUE TITLE
question about data processing

### ISSUE THREAD

**Issue by huzhen-ah** [nastassiavysotskaya/BP_Transformer#1]
Hello, I have a data processing question. Each subject has N segments, but these N segments' segment_ids are not continuous. For example, 5, 4, 3, 19, 18, 17, 39, 38, 37. When creating the dataset, if each segment generates features of (10, 21), I noticed in buildFeatureDatabaseAbp.py that the last segment is discarded. Therefore, the total features would be (80, 21). During the dataloader process, why do we slide over the time dimension of 90? To me, since the segment_ids are neither continuous nor ordered, the time sequence is messed up, so how can it be slid?

**Comment by nastassiavysotskaya** [comment]
Hello!
Did you try setting shuffling to False?

**Comment by huzhen-ah** [comment]
I saw PulseDB on a GitHub project, which includes code for generating Subset_Files. The code is as follows:
《https://github.com/pulselabteam/PulseDB/blob/main/Generate_Subsets.m》
Among which there is a section:

    Segments_File=load(fullfile(Segment_Path,Subj_ID));
    Subj_Segments=Segments_File.Subj_Wins;
    Selected_IDX=[Info(strcmp({Info.Subj_Name},Subj_Name)).Subj_SegIDX]; % All selected segments belonging to this subject
    for j=Selected_IDX
        Segment=Subj_Segments(j);
        Subset.Subject{pos}=Subj_Name;
        Subset.Signals(pos,:,:)=[Segment.ECG_F,Segment.PPG_F,Segment.ABP_Raw]';
        Subset.SBP(pos)=Segment.SegSBP;
        Subset.DBP(pos)=Segment.SegDBP;
        Subset.Age(pos)=Segment.Age;
        Subset.Gender{pos}=Segment.Gender;
        if Source==1 %Record information for VitalDB subjects
            Subset.Height(pos)=Segment.Height;
            Subset.Weight(pos)=Segment.Weight;
            Subset.BMI(pos)=Segment.BMI;
        end
        pos=pos+1;
    end  


This section of code does not take into account the temporal sequence of segment_id, so the N segment signals corresponding to the same subject_id in the generated file, such as Train_Subset.mat, are also not in chronological order. For example, they might be: segment_1_signals, segment_3_signals, segment_5_signals, segment_2_signals. When using buildFeatureDatabaseAbp.py to generate the dataset in your project, the temporal order between the aforementioned segments was also not considered. Later, I noticed that in the dataloader.py, the features generated from all segments under the same segment_id were grouped together (without considering the temporal order of the segments), and then sliding sampling was performed along the time dimension for training. 
some code in your dataloader.py



    def preprocess_data(self, dataframeWithNaN):
        ...some code...
        target_dict[subject_id] = target  # torch.FloatTensor(target).to(device)
        demographic_dict[subject_id] = demographic  # torch.FloatTensor(demographic).to(device) #.iloc[0]
        data_dict[subject_id] = data  # torch.FloatTensor(data).to(device)
        ...some code...

    def __getitem__(self, idx):
            # self.__len__
            start_idx = idx
            for subject_ID in self.subject_ids:
                # print(idx, start_idx, len(self.data[subject_ID]))
                if start_idx < len(self.data[subject_ID]) - self.sequence_length + 1:
                    data_from_subject = self.data[subject_ID][start_idx:start_idx + self.sequence_length - 1]
                    data_from_subject = torch.Tensor(data_from_subject).to(self.device)
                ....some code....
            ...some code.....


I believe this is unreasonable. Since sliding sampling across segments is being carried out, I think all segments should be in order. For example, segment_1, segment_2, segment_3 is acceptable, but segment_1, segment_3, segment_2 or segment_3, segment_2, segment_1 or segment_3, segment_1 are not acceptable.
